src/Pos/Genesis.hs

```diff
@@ -141,37 +89,144 @@ bitcoinDistributionImpl ratio coins (coinIdx, coin) =
+        x `unsafeMulCoin` coinToInteger (coins `divCoin` (1000 :: Int))
+stakeDistribution ExponentialStakes = map (, []) expTwoDistribution
+stakeDistribution ts@RichPoorStakes {..} =
+    checkMpcThd (getTotalStake ts) sdRichStake $
+    map (, []) basicDist
+  where
+    -- Node won't start if richmen cannot participate in MPC
+    checkMpcThd total richs =
+        if richs < applyCoinPortion Const.genesisMpcThd total
+        then error "Pos.Genesis: RichPoorStakes: richmen stake \
+                   \is less than MPC threshold"
+        else identity
+    basicDist = genericReplicate sdRichmen sdRichStake ++
+                genericReplicate sdPoor sdPoorStake
+stakeDistribution (CustomStakes coins) = map (,[]) coins
+
+-- | Generates genesis 'Utxo' given optional boot stakeholders list
+-- and address distribution. In case boot stakeholders are supplied,
+-- all the balance is distributed among them. Otherwise, txDistr is
+-- set to @[]@ or left as it is (in case of 'ExplicitStakes').
+genesisUtxo :: Maybe (HashSet StakeholderId) -> [AddrDistribution] -> Utxo
+genesisUtxo bootStakeholders ad =
+    M.fromList $ concatMap (uncurry toUtxo . second stakeDistribution) ad
+  where
+    defaultStakeDistr coin distr
+        = maybe distr
+                (\genDistr -> genesisSplitBoot genDistr coin)
+                bootStakeholders
+    toUtxo :: [Address] -> [(Coin, TxOutDistribution)] -> [(TxIn, TxOutAux)]
+    toUtxo addr distrs =
+        let utxoEntry a (coin,txOutDistr) =
+                ( TxIn (unsafeHash a) 0
+                , TxOutAux (TxOut a coin) (defaultStakeDistr coin txOutDistr)
+                )
+        in zipWith utxoEntry addr distrs
+
+-- This is probably 100% useless.
 genesisDelegation :: HashMap StakeholderId (HashSet StakeholderId)
 genesisDelegation = mempty
 
+-- | Compute leaders of the 0-th epoch from stake distribution.
+genesisLeaders :: Utxo -> SlotLeaders
+genesisLeaders utxo =
+    followTheSatoshi genesisSeed $ HM.toList $ utxoToStakes utxo
+
 ----------------------------------------------------------------------------
--- Slot leaders
+-- Development mode genesis
 ----------------------------------------------------------------------------
 
--- | Compute leaders of the 0-th epoch from stake distribution.
-genesisLeaders :: StakesMap -> SlotLeaders
-genesisLeaders genesisStakes =
-    followTheSatoshi genesisSeed $ HM.toList genesisStakes
+-- | First index in derivation path for HD account, which is put to genesis utxo
+accountGenesisIndex :: Word32
+accountGenesisIndex = firstNonHardened
+
+-- | Second index in derivation path for HD account, which is put to genesis
+-- utxo
+wAddressGenesisIndex :: Word32
+wAddressGenesisIndex = firstNonHardened
+
+-- | Chooses among common distributions for dev mode.
+devStakesDistr
+    :: Maybe (Int, Int)                   -- flat distr
+    -> Maybe (Int, Int)                   -- bitcoin distr
+    -> Maybe (Int, Int, Integer, Double)  -- rich/poor distr
```

Comment:
I want to merge it asap and i think it's minor. Moreover it is here since the beginning of time, i've just moved function around.